### PR TITLE
handle namespaceselector and empty namespaces

### PR DESCRIPTION
**Pull request by ritazh**
Signed-off-by: Rita Zhang <rita.z.zhang@gmail.com>

Fixes https://github.com/open-policy-agent/gatekeeper/issues/153

### CODE REVIEW

File: constraint/pkg/client/regolib/src.go [hunk 1]
```diff
@@ -4,10 +4,24 @@ const (
 	targetLibSrc = `
 package hooks["{{.Target}}"]
 
+# Skip constraint evaluation for a given target if constraint has namespaceSelector but Namespace is not cached
```
Comment (maxsmythe) [inline constraint/pkg/client/regolib/src.go]: k8s-specific code should not be in the framework. The framework should provide a way for the target-handler to auto-reject and put the NS selector auto-rejection in the targethandler

Comment (ritazh) [reply constraint/pkg/client/regolib/src.go]: How about adding a new rule similar to `matching_constraints[constraint]` in `Library()` ? For example, the new rule could be called ` autoreject_constraints[constraint, review]` and implemented by the target handler.

Comment (maxsmythe) [reply constraint/pkg/client/regolib/src.go]: Per our conv. on Friday, that SGTM

File: constraint/pkg/client/regolib/src.go [hunk 2]
```diff
@@ -4,8 +4,14 @@ const (
...
 package hooks["{{.Target}}"]
 
+# Auto reject reviews when a constraint has namespaceSelector but Namespace is not cached
+deny[response] {
+	data.hooks["{{.Target}}"].library.autoreject_review[response]
+}
+
 # Finds all violations for a given target
 deny[response] {
+	not data.hooks["{{.Target}}"].library.autoreject_review[response]
```
Comment (maxsmythe) [inline constraint/pkg/client/regolib/src.go]: this `not` line should be dropped. We should keep reporting violations even if autoreject is also true as the intent is to return as much actionable feedback (violations) as possible.

File: constraint/pkg/client/regolib/src.go [hunk 3]
```diff
@@ -4,8 +4,14 @@ const (
 	targetLibSrc = `
 package hooks["{{.Target}}"]
 
+# Auto reject reviews when a constraint has namespaceSelector but Namespace is not cached
+deny[response] {
```
Comment (maxsmythe) [inline constraint/pkg/client/regolib/src.go]: The format for `response` should be handled by this Rego package, as the CF relies on the interface defined by response to parse/return violations.

**Comment by maxsmythe** [comment]
Somehow didn't see that `regolib/src.go` was edited. Dunno why. Sorry :/

File: constraint/pkg/client/regolib/src.go (hunk 3 above)
Comment (ritazh) [reply constraint/pkg/client/regolib/src.go]: so maybe `autoreject_review` should just return the msg?

File: constraint/pkg/client/regolib/src.go [hunk 4]
```diff
@@ -4,6 +4,17 @@ const (
 	targetLibSrc = `
 package hooks["{{.Target}}"]
 
+violation[response] {
+	data.hooks["{{.Target}}"].library.autoreject_review[[rejection, constraint]]
```
Comment (maxsmythe) [inline constraint/pkg/client/regolib/src.go]: `constraint` should be a property of `rejection`. This way handlers that reject for reasons that have nothing to do with the constraint don't need to worry about creating a dummy constraint.

File: constraint/pkg/client/regolib/src.go [hunk 5]
```diff
@@ -4,6 +4,17 @@ const (
...
+	review := get_default(input, "review", {})
+	response = {
```
Comment (maxsmythe) [inline constraint/pkg/client/regolib/src.go]: We should test that these results actually get passed on to the client. Unit tests & e2e tests.

**Pull request merged**